### PR TITLE
Development

### PR DESCRIPTION
**Pull request by dipto0321**
Hi,
App features:
1. Created a default project if there is no project.
2. A user can add todos in the default project
3. Can add new project and to-dos according to that project.
4. A user can update the project and delete the project.
5. Show todos by click on the project list item.
6. Todos can be updated and deleted. Also can mark the todo complete.

We set it up based on MVC pattern (though we don't have any views). We separated classes (project and todo) from the logic (which we put in the controllers). We also have separate todos and projects helper modules. We try to follow the **core** instructions from the Odin and from the mentor guideline. While, we make sure that we satisfy all the requirements/specs in the project and also considering the context of the topic under which this project falls, we don't follow the instructions line by line as there may be some steps that are redundant/unnecessary. So for code reviewers please consider checking the general specification of the project and check them against what we did and not too much on the details of the steps. Thanks

### CODE REVIEW

**Comment by talihomz** [comment]
## Awesome ToDo List App!
---

Hi @dipto0321 @rvvergara. 
This looks good, was able to add projects and assign tasks to them. Was abit hard figuring out how to add tasks though (had to press enter, haha), but everything works fine!
Great project structure as well! I'l leave comments on specific files where I think improvements can be done.

File: src/controllers/helpers/projects/projectHelpers.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+} from '../todos/todoBody';
+import projectsController from "../../projectsController";
+
+import {
+  generateAddTodoBtn
+} from "../todos/todoHelpers";
+
+export const showProjectList = () => {
+  // show project list only if there are projects in the projectsArray
+  let projectsArray = JSON.parse(localStorage.getItem("projectsArray"));
```
Comment (talihomz) [inline src/controllers/helpers/projects/projectHelpers.js]: This DOM helper should not have any idea of where the data is stored! (localStorage). Move this functionality (`localStorage.getItem("projectsArray")`) to its own file (probably `storage.js`)

File: src/controllers/helpers/projects/projectHelpers.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+  span.innerText = project.name;
+
+  // Give each li a class of list-group-item and id based on project's id
+  li.setAttribute("class", "list-group-item");
+  li.setAttribute("id", `projectLi-${project.id}`);
+
+  // Add click event listener to each li
+  li.addEventListener('click', () => {
+    // Override the existing project with updated project
+    project = JSON.parse(localStorage.projectsArray).find(x => x.id === project.id);
```
Comment (talihomz) [inline src/controllers/helpers/projects/projectHelpers.js]: Same as previous comment. This helper should not access the DB implementation, rather an abstraction of it.

File: src/controllers/helpers/projects/projectHelpers.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  return btn;
+};
+
+// Adding event listeners to btns
+const addBtnEventListeners = (btn, project, action) => {
+
+  btn.addEventListener("click", e => {
+    e.stopPropagation();
+    // Override the existing project with updated project
+    project = JSON.parse(localStorage.projectsArray).find(x => x.id === project.id);
```
Comment (talihomz) [inline src/controllers/helpers/projects/projectHelpers.js]: Same as above. Move to `storage.js`.

File: src/controllers/helpers/projects/projectHelpers.js [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    document.getElementsByTagName("ul")[0].appendChild(li);
+  }
+};
+
+// Creating a projects array for new users or after localstorage is cleared
+
+export const createProjectsArray = () => {
+  if (localStorage.getItem("projectsArray") === null) {
+    let projectsArray = [];
+    localStorage.setItem("projectsArray", JSON.stringify(projectsArray));
```
Comment (talihomz) [inline src/controllers/helpers/projects/projectHelpers.js]: `storage.js`

File: src/controllers/helpers/projects/projectHelpers.js [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  }
+};
+
+// Create a default project for new users or after localstorage is cleared
+export const createDefaultProject = () => {
+  if (localStorage["projectCount"] === "0") {
+    let defaultProject = projectsController.create("Default Project");
+  }
+
+  if (localStorage["Default Project"] !== undefined) {
```
Comment (talihomz) [inline src/controllers/helpers/projects/projectHelpers.js]: `storage.js`

File: src/controllers/helpers/todos/todoBody.js [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    // Iterate thru each todo and create tr for them
+    project.todos.forEach(todo => {
+      createTodoRow(todoBody, todo, project);
+    });
+  }
+};
+
+// Get project name from  localstorage projectsArray
+export const extractProjectName = dataID => {
+  return JSON.parse(localStorage["projectsArray"]).find(x => x.id === dataID).name;
```
Comment (talihomz) [inline src/controllers/helpers/todos/todoBody.js]: `storage.js`

File: src/controllers/helpers/todos/todoBody.js [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+};
+
+// Get project from localstorage
+export const extractProject = name => {
+  return JSON.parse(localStorage[name]);
```
Comment (talihomz) [inline src/controllers/helpers/todos/todoBody.js]: `storage.js`

File: src/controllers/helpers/todos/todoHelpers.js [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  if (document.getElementsByClassName("addTodoBtn")[0]) {
+    document.getElementById("todosDiv").removeChild(document.getElementsByClassName("addTodoBtn")[0]);
+  }
+  if (document.getElementById(`addTodoBtn-${project.id}`) === null) {
+    document.getElementById("todosDiv").appendChild(btn);
+  }
+};
+
+export const submitTodoCallBack = target => {
+  let projectsArray = JSON.parse(localStorage["projectsArray"]),
```
Comment (talihomz) [inline src/controllers/helpers/todos/todoHelpers.js]: `storage.js`

File: src/controllers/projectsController.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
+import Project from '../models/project';
+
+const projectsController = (
+  () => {
+    let projectsArray = JSON.parse(localStorage.getItem("projectsArray")) ? JSON.parse(localStorage.getItem("projectsArray")) : [];
```
Comment (talihomz) [inline src/controllers/projectsController.js]: `storage.js`

File: src/controllers/projectsController.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    return {
+      create(name) {
+        if (name === "") name = "Project Name (Please customize name)";
+        let project = new Project(name);
+        // localStorage.setItem(name, JSON.stringify(project));
+        saveToLocalStorage(name, project);
```
Comment (talihomz) [inline src/controllers/projectsController.js]: Where is `saveToLocalStorage` defined? There are no imports to it, so i'm assuming its a global?

File: src/controllers/projectsController.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+        projectsArray.push(project);
+        saveToLocalStorage("projectsArray", projectsArray);
+        return project;
+      },
+      update(id) {
+        let oldProjectName = projectsArray.find(x => x.id === id).name;
+        let oldProject = JSON.parse(localStorage[oldProjectName]);
+        let project = JSON.parse(localStorage[oldProject.name]);
```
Comment (talihomz) [inline src/controllers/projectsController.js]: `storage.js`

File: src/controllers/projectsController.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        let index = projectsArray.findIndex(x => x.name === project.name);
+        localStorage.removeItem(name);
+        projectsArray.splice(index, 1);
+        saveToLocalStorage("projectsArray", projectsArray);
+      }
+    };
+  }
+)();
+
+const saveToLocalStorage = (name, value) => {
```
Comment (talihomz) [inline src/controllers/projectsController.js]: Aaah, here it is. It doesn't belong here though! Move this to `storage.js`

File: src/controllers/todosController.js [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+const todoController = (
+  () => {
+    return {
+      create(title, description, dueDate, priority, notes = "", project) {
+        let todo = new Todo(title, description, dueDate, priority, notes, project);
+        // Extract parent project from localStorage
+        let parentProject = JSON.parse(localStorage.getItem(project));
+        // Push todo into parentProj's todo
+        parentProject.todos.push(todo);
+        localStorage.setItem(project, JSON.stringify(parentProject));
```
Comment (talihomz) [inline src/controllers/todosController.js]: `storage.js`

File: src/controllers/todosController.js [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+          description,
+          dueDate,
+          priority,
+          notes,
+          project: name,
+          done: done,
+          id: id
+        };
+        project.todos.splice(index, 1, todoUpdated);
+        localStorage.setItem(project.name, JSON.stringify(project));
```
Comment (talihomz) [inline src/controllers/todosController.js]: `storage.js`

File: src/controllers/todosController.js [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+      },
+      delete(project, id) {
+        // Extract parent project
+        let parentProject = JSON.parse(localStorage.getItem(project));
+        // Find todo from name
+        let todoIndex = parentProject.todos.findIndex(x => x.id == id);
+        // Splice project
+        parentProject.todos.splice(todoIndex, 1);
+        localStorage.setItem(project, JSON.stringify(parentProject));
```
Comment (talihomz) [inline src/controllers/todosController.js]: `storage.js`

File: src/index.js [hunk 16]
```diff
@@ -1 +1,39 @@
...
+  submitTodoCallBack,
+} from './controllers/helpers/todos/todoHelpers';
+
+createProjectsArray();
+createDefaultProject();
+showProjectList();
+
+// For refactor:
+// Adding event listener to Project Form
+document.getElementById("projectNameForm").addEventListener('submit', e => {
```
Comment (talihomz) [inline src/index.js]: The `index.js` is probably where all the magic for ToDo List App happens. These DOM listeners need to be in their own file, like you did for the task and project helpers

File: src/models/project.js [hunk 17]
```diff
@@ -0,0 +1,13 @@
+export default class Project {
+  constructor(name) {
+    this.name = name;
+    this.todos = [];
+    let projectCount = Number(localStorage.projectCount);
+    this.id = ++projectCount;
+    localStorage.setItem("projectCount", projectCount);
```
Comment (talihomz) [inline src/models/project.js]: The model `Project` should not be aware about the underlying storage, nor how projects are incremented! It should strictly create a new project model, and return it. 

Incrementing the projects count is probably a responsibility of the Projects Controller.

File: src/models/project.js [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+  }
+}
+
+if (localStorage["projectCount"] === undefined) {
+  localStorage.setItem("projectCount", 0);
```
Comment (talihomz) [inline src/models/project.js]: `storage.js`

File: src/models/todo.js [hunk 19]
```diff
@@ -0,0 +1,12 @@
+export default class Todo {
```
Comment (talihomz) [inline src/models/todo.js:1]: This class was perfectly done! It only does its job, creating a ToDo and returning it. Use this way of thinking for the Project class as well.

**Review by talihomz — CHANGES_REQUESTED** [review]

**Comment by rvvergara** [comment]
Awesome @talihomz! Thanks for the concise yet very clear review. We love how you put context into what you suggest for us to do :) We'll work on those changes

**Comment by dipto0321** [comment]
Thanks @talihomz ,
This is called **NEAT and CLEAN** feedback 😸

File: src/controllers/helpers/projects/projectHelpers.js (hunk 1 above)
Comment (dipto0321) [reply src/controllers/helpers/projects/projectHelpers.js]: Solved

File: src/controllers/helpers/projects/projectHelpers.js (hunk 2 above)
Comment (dipto0321) [reply src/controllers/helpers/projects/projectHelpers.js]: solved

File: src/controllers/projectsController.js (hunk 9 above)
Comment (dipto0321) [reply src/controllers/projectsController.js]: solved

File: src/index.js (hunk 16 above)
Comment (dipto0321) [reply src/index.js]: Fixed

**Comment by talihomz** [comment]
## Guys... 👏👏
---

This project is really clean and neat guys!!

👍

**Pull request merged**